### ISSUE TITLE
OpenId configuration problem

### ISSUE THREAD

**Issue by yebenes** [LuatixHQ/opencti#3718]
## Prerequisites

- [X] I read the [Deployment and Setup](https://www.notion.so/OpenCTI-Public-Knowledge-Base-d411e5e477734c59887dad3649f20518) section of the OpenCTI documentation as well as the [Troubleshooting](https://www.notion.so/Troubleshooting-ebc8fb04137d495aad917bc20340b9a6) page and didn't find anything relevant to my problem.
- [X] I went through old GitHub issues and couldn't find anything relevant
- [X] I googled the issue and didn't find anything relevant

## Description

I'm trying to configure OpenID in OpenCTI version 5.9.0 but I can't get it to work. I think the problem is related with the groups path use to get the groups.

These are the environment variables I have define:

`
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__ORGANIZATIONS_PATH=["roles"]
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__GROUPS_PATH=["roles"]
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__ORGANIZATIONS_MAPPING=["TEST-ADMIN:EUROPE"]
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__GROUPS_MAPPING=["TEST-ADMIN:Administrator"]
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__GROUPS_SCOPE=roles
PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__ORGANIZATIONS_SCOPE=roles
`

And the logs from OpenId are:

`
{"category":"APP","level":"debug","message":"[OPENID] Successfully logged","timestamp":"2023-07-11T10:59:19.592Z","userinfo":{"country":"ESP","email":"hello_user@test.com","given_name":"Hello","name":"Hello User","preferred_username":"XXXXXX","roles":["TEST-ADMIN"],"sub":"XXXXXX"},"version":"5.9.0"}
`

 I think like the param with mi roles is "roles" and the PROVIDERS__OPENID__CONFIG__ROLES_MANAGEMENT__GROUPS_PATH define ["roles"] as path, it should work but when I enter with my user I see I have the default role.

Could you help me? I don't know how can I fix it.

Thanks a bunch,
Alberto.

**Comment by SamuelHassine** [comment]
Hello @yebenes,

Please be aware of the fact that the role is case sensitive, on our side, this configuration is working:

```
"groups_management": {
  "groups_scope": "roles",  
  "groups_path": ["roles", "realm_access.roles", "resource_access.account.roles"],
  "groups_mapping": ["administrator:Administrators", "user:Default"]
}
``` 

where administrator / user are roles, and Administrators / Default are groups in OpenCTI.

**Comment by SamuelHassine** [comment]
Also, for environment variables with list of strings, please refer to the documentation:

```
- "PROVIDERS__OPENID__CONFIG__GROUPS_MANAGEMENT__GROUPS_MAPPING=[\"OpenID_Group_1:OpenCTI_Group_1\", \"OpenID_Group_2:OpenCTI_Group_2\", ...]"
```

**Comment by SamuelHassine** [comment]
Your need double quotes around the variable and escape inside with backslash.

https://docs.opencti.io/latest/deployment/authentication/#openid-connect-button

**Comment by yebenes** [comment]
Hi Samuel,

Yes, you're right, my current configuration is:

```
  # OPENID:
  PROVIDERS__OPENID__STRATEGY: "OpenIDConnectStrategy"
  PROVIDERS__OPENID__CONFIG__LABEL: "Login with IDP"
  PROVIDERS__OPENID__CONFIG__ISSUER: "https://idp.****"
  PROVIDERS__OPENID__CONFIG__REDIRECT_URIS: "[\"https://opencti.url/auth/oic/callback\"]"

  # OPENID GROUPS ROLE MAPPING:
  PROVIDERS__OPENID__CONFIG__GROUPS_MANAGEMENT__GROUPS_PATH: "[\"roles\"]"
  PROVIDERS__OPENID__CONFIG__GROUPS_MANAGEMENT__GROUPS_MAPPING: "[\"ACME_EU-ADMIN:Administrator\",\"ACME_HO-ADMIN:Administrator\",\"ACME_HO-ANALYST_CTI:Analyst\",\"ACME_HO-VIEWER_CTI:Viewer\"]"

  # OPENID ORGS ROLE MAPPING:
  PROVIDERS__OPENID__CONFIG__ORGANIZATIONS_MANAGEMENT__ORGANIZATIONS_PATH: "[\"roles\"]"
  PROVIDERS__OPENID__CONFIG__ORGANIZATIONS_MANAGEMENT__ORGANIZATIONS_MAPPING: "[\"ACME_EU-ADMIN:ACME_EU\",\"ACME_HO-ADMIN:ACME_HO\",\"ACME_HO-ANALYST_CTI:ACME_HO\",\"ACME_HO-VIEWER_CTI:ACME_HO\"]"

```

But it still doesn't work, this morning you told me that there was a problem with the OpenId login in version 5.9.0.

Do I need to wait for it to be fixed?

Thanks a bunch,
Alberto.

**Comment by SamuelHassine** [comment]
Hello Alberto,

After careful troubleshooting on our side and multiple testing, everything is working fine with OIDC for all customers and ourselves.

Can you please share the log with something like (debug mode):

![image](https://github.com/OpenCTI-Platform/opencti/assets/1334279/86966c02-c9f3-4ba1-ada6-0166e67823d6)

You should check if mapping is available in the OIDC answers (aka roles) and if groups / organizations exist in the OpenCTI platform.

Kind regards,
Samuel

**Comment by yebenes** [comment]
Hi Samuel,

How are you? First, thanks for your quickly response. 

Here are the logs of my application:

```
{
    "category": "APP",
    "level": "debug",
    "message": "[OPENID] Successfully logged",
    "timestamp": "2023-07-19T06:55:50.922Z",
    "userinfo": {
        "country": "ESP",
        "email": "john.doe@company.com",
        "given_name": "JOHN",
        "name": "JOHN DOE",
        "preferred_username": "john",
        "roles": ["ADMIN"],
        "sub": "john"
    },
    "version": "5.9.0"
} 
{
    "category": "APP",
    "decoded": {
        "azp": "fa89d7b3-b297-4a57-8f39-807123cbbcf7",
        "exp": 1689792950,
        "iat": 1689749750,
        "iss": "https://idp.company.com/oidc/",
        "jti": "47502114-5189-4c48-a70e-7f511a51dbaf",
        "sub": "john"
    },
    "level": "debug",
    "message": "[OPENID] Groups mapping on decoded access_token",
    "timestamp": "2023-07-19T06:55:50.923Z",
    "version": "5.9.0"
}
{
    "category": "APP",
    "error": "Restricted access, ask your administrator",
    "level": "error",
    "message": "Error login through provider oic",
    "timestamp": "2023-07-19T06:55:50.925Z",
    "version": "5.9.0"
}
```
I have a theory that I don't know if you can validate me.

In the first log, with the message [OPENID] Successfully logged, we see that the login was correct and my user's information is displayed in the userinfo field. However, this information is not used for role mapping.

After this call, the access_token is decoded and tries to use the fields that are included in it, but the roles are not found in this token.

At this point I have a questions, could the first token be used for group mapping?

Thanks a bunch for your help.

Best,
Alberto.

**Comment by yebenes** [comment]
Hi @SamuelHassine,

Can you think of anything else I can look at? I don't know how we could solve this.

Thanks a bunch,
Alberto.

**Comment by CyberKaizen** [comment]
@yebenes I've been having this Groups/Organization mapping issue for 7 months. I've asked many in the community for assistance as well. But nobody seems to have gotten this mapping to work with OpenID or SAML. I've even tested this functionality with KeyCloak which seems to be what the devs will use to test different configurations with no luck.

The workaround I've done is:
- Setting a default group in the platform that users automatically join
- Using this environment variable to set a default org for users to join when they first login via SSO: `PROVIDERS__OPENID__CONFIG__ORGANIZATIONS_DEFAULT=[\"YourOrgHere\"]"`
- If you have a platform org set, you will need to use the Organization that is your platform org in that setting or you'll get an error that won't specifically let you know about this problem.
- Then I manually move users to the appropriate group or organization.

**Note: Having a platform organization set can really cause you issues when troubleshooting SSO integration. I recommend removing it while troubleshooting.**

Hope this helps!

**Comment by yebenes** [comment]
Hi @CyberKaizen,

I resolved my problem, I don't know if my solution could help you.

By default, OpenCTI uses the access_token to get the organization's information. In my case, my access_token doesn't contain the organization's information.

I need to set the TOKEN_REFERENCE param to "id_token" to get this information from the id_token.

The configuration that I defined is:
 
`PROVIDERS__OPENID__CONFIG__ORGANIZATIONS_MANAGEMENT__TOKEN_REFERENCE: "id_token"`

Best,
Alberto.
